sql/sql_acl.cc

```diff
@@ -10551,6 +10637,91 @@ int mysql_alter_user(THD* thd, List<LEX_USER> &users_list)
   DBUG_RETURN(result);
 }
 
+/*
+  Mark an user account as locked or unlocked.
+
+  SYNOPSIS
+    lock_user_account()
+    thd                         The current thread.
+    user_table                  A TL_WRITE opened User_table
+    lex_user                    The user to lock/unlock.
+    lock_cmd                    true locks an user, false unlocks.
+
+  RETURN
+    > 0         Error. Error message already sent.
+    0           OK.
+*/
+static int lock_user_account(THD* thd, const User_table &user_table,
+                             LEX_USER *lex_user, bool lock_cmd)
+{
+  char user_key[MAX_KEY_LENGTH];
+  int error;
+
+  DBUG_ENTER("lock_user_account");
+
+  mysql_mutex_assert_owner(&acl_cache->lock);
+
+  TABLE *table= user_table.table();
+
+  /*
+     Do not allow the current user to lock itself out.
+  */
+  ACL_USER *current_acl_user= find_user_wild(thd->security_ctx->priv_host,
+                                             thd->security_ctx->priv_user);
+  if (!current_acl_user ||
+      (!strcmp(lex_user->user.str, current_acl_user->user.str) &&
+       !my_strcasecmp(system_charset_info, lex_user->host.str,
+                      current_acl_user->host.hostname)))
+  {
+    DBUG_RETURN(1);
+  }
+
+  const char *host= lex_user->host.str;
+  const char *username= lex_user->user.str;
+  ACL_USER *acl_user;
+  if (!(acl_user= find_user_exact(host, username)))
+  {
+    my_error(ER_PASSWORD_NO_MATCH, MYF(0));
+    DBUG_RETURN(1);
+  }
+
+  /*
+     Move to the next user if the current one is already locked/unlocked
+  */
+  if (acl_user->account_locked == lock_cmd)
```

Comment:
This, in theory is a nice optimization. Here's the problem:

User does the following:
alter user foo account locked;
update mysql.user (or json equiv) set mysql.account_locked=N;
<forgets to run flush privileges>
alter user foo account locked; # This won't change the table data.
<user reports bug>
We waste a bunch of hours figuring out what the problem is.

As this is not performance critical code, I'd just skip this bit. I couldn't find similar logic within sql_acl, to at least show a precedent.